### ISSUE TITLE
QTable with virtual scrolling loses border-bottom with many rows

### ISSUE THREAD

**Issue by delphi-sucks** [quasarframework/quasar#17566]
### What happened?

If one uses a QTable with virtual scrolling and many rows (> 400.000), than the rows lose the `border-bottm` style after a certain amount of rows when scrolling.

### What did you expect to happen?

Every row, regardless of its position, should have the same style and therefore a `border-bottom` in this case.

### Reproduction URL

https://codepen.io/delphi-sucks/pen/oNKzKjj

### How to reproduce?

1. Create a QTable with virtual scrolling
2. Have at least ~400.000 rows
3. Scroll to the end
=> Rows don't have a `border-bottom` anymore compared the the first rows.

### Flavour

Vite Plugin (@quasar/vite-plugin)

### Areas

Components (quasar)

### Platforms/Browsers

Chrome

### Quasar info output

```Shell
Operating System                        Windows_NT(10.0.22631) - win32/x64
NodeJs                                  20.17.0

Global packages
  NPM                                   10.8.3
  yarn                                  Not installed
  pnpm                                  Not installed
  bun                                   Not installed
  @quasar/cli                           2.4.1
  @quasar/icongenie                     Not installed
  cordova                               Not installed
```

### Additional context

Chrome 129.0.6668.90

**Comment by RamonDonadeu** [comment]
Adding more info:

It seems like border-bottom-width is treating 0px and 1px as 0px, in this case inspecting the DOM I can see a very tiny border, but it is not visible, the same happens with:

3px 4px 5px -> treated as 4px
7, 8 and 9px as 8px
11px, 12px and 13px as 12px

In fact, looking closely, I can see that, in 1px, 3px, 7px and 11px, the border viewed is like 0.2px bigger than what DOM wants to render. 

![Captura de pantalla 2024-10-10 a las 9 27 51](https://github.com/user-attachments/assets/f05b7c87-9cd0-4c38-b2f8-6f0bd0441427)
You can see that there is a black border overflowing the orange block which is the border.

Then at 4px, 7px, 13px, it is the desired border
And the in 5px, 9px, 13px, the border to render is bigger, than what is rendered 
![Captura de pantalla 2024-10-10 a las 9 25 58](https://github.com/user-attachments/assets/48c0a4f8-921c-4674-86c2-e1df22fa880d)
And here there is a more yellowish part that should be filled black.

Also, the last row to have border-bottom is 349524, even if there are 500000 rows, this behavior starts at the same place.
When you are below row 349525, I can't scroll anymore, it is blocked. I have to scroll two times to unblocking it? (Tried with magic moude) This is so weird.

**Comment by rickxu0423** [comment]
Encountered similar issue here, if you set a height for the table and have very few rows (eg: height: 400px with 2 rows), the border-bottom is missing from for the very last row. https://codepen.io/Rickxu0423/pen/azogGKb

**Comment by FangHanZreo** [comment]
> 在这里遇到了类似的问题，如果您为表格设置了高度并且行数很少（例如：height： 400px 有 2 行），则最后一行缺少 border-bottom. https://codepen.io/Rickxu0423/pen/azogGKb

This is not a bug, but rather an issue where the last row does not have a border-bottom set. When the table is fully populated (stretched to fill its container), the border of the last row overlaps with the border-top of q-table__bottom. This results in the following visual effect:
![Image](https://github.com/user-attachments/assets/8758eb8c-587a-49e2-8921-0375948f31a9)

**Comment by FangHanZreo** [comment]
I am also experiencing an issue with table borders not displaying correctly. [Reproduction link](https://codepen.io/fanghanzreo/pen/PwoVWpL?editors=111).

![Image](https://github.com/user-attachments/assets/2e36ff83-23e3-46d3-89e3-bfedaccdc441)

**Reproduction Conditions:**
1. Use position: sticky to fix columns.
2. Make the content inside q-td relatively tall (40px).
3. Zoom the webpage to set window.devicePixelRatio to 1.25.

**Solution:**
Adjusting the default line height of q-td can resolve this issue.

**Personal Analysis:**
This appears to be a browser rendering issue.

**Comment by rstoenescu** [comment]
Thanks for the report, and thanks @RamonDonadeu for the measurements, they turned out to be the key to it.

This is a browser paint limitation rather than something Quasar computes wrong, so I'm closing it.

### Root cause

Chromium stops painting at a document/scroll offset of 2^24 = 16,777,216 CSS px. Below that everything is fine; at and above it, hairline geometry is simply not painted. Layout is unaffected: `getBoundingClientRect()` returns the right values, the computed `border-bottom-width` still reports `1px`, cell text still renders, and `scrollTop` keeps accepting values up to `scrollHeight`. It is float32 mantissa exhaustion (24 bits) in Blink's CSS space geometry, not device space, since the threshold is identical at `devicePixelRatio` 1 and 2.

I rendered a real QTable with 500,000 rows at the default 48px row height and screenshotted it: the last row with a visible border is 349524, exactly the number @RamonDonadeu measured. 349524 * 48 = 16,777,152, so that border paints at 16,777,200, just under 2^24, while row 349525's border would land at 16,777,248, just over.

The border width quantization is the same effect. Above 2^24 the float32 ULP is 2, so at Y = 20,000,000 a 1px border vanishes, 2px stays 2px, and 3px/4px/5px all render as 4px, matching your measurements exactly.

Other engines have their own ceilings. Last offset at which a 1px border still paints, binary searched:

| Engine | Last painting offset | Fails from | Why |
| --- | --- | --- | --- |
| Chromium | 16,777,152 | 16,777,200 | 2^24, float32 mantissa |
| Firefox | 17,895,648 | 17,895,696 | `nscoord_MAX/2` = 2^31/(2*60), Gecko's 60 app units per px on int32. Past it nothing paints at all, no border width helps |
| WebKit | at least 32,000,000 (my test ceiling) | not reached | 1px to 5px all exact |

### What this means in practice

The usable row count on Chromium is 2^24 / rowHeight:

- ~349,525 rows at the default 48px
- ~524,288 rows with the `dense` prop (32px)
- ~699,050 rows at 24px

So `dense` buys you roughly 50% more rows. Beyond that, the only way out is to keep the whole dataset out of a single scroll space: server side pagination, or filtering/grouping the list down.

Supporting 400k+ rows on Chromium would mean compressing the scroll space, capping the total spacer height and mapping `scrollTop` onto the row range non linearly, the way AG Grid does. Our virtual scroll assumes a 1:1 pixel mapping throughout, so that is a substantial rework for a case the platform cannot paint anyway. Not something we plan to take on, but it is written down.

### On the other reports in this thread

@rickxu0423 @FangHanZreo, those are separate matters and are not covered by the above:

- The missing border under the last row of a short table is intentional: the separator rule is `tbody > tr:not(:last-child) > td`, and `q-table__bottom`'s `border-top` normally supplies that line. @FangHanZreo's explanation is correct.
- The seams with sticky columns at `devicePixelRatio` 1.25 are fractional DPR pixel snapping in the browser. If you still see it on a current version, please open a separate issue with a reproduction.
